### PR TITLE
Fix/code reviews

### PR DESCRIPTION
**Pull request by Mahmoud-AbdElaziz-Hamed**

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -3,57 +3,57 @@ import { useEffect, useState } from "react";
...
   const handleInputChange = (e) => {
     setAddValue(e.target.value);
   };
-  const handleAdd = (value) => {
-    setData([...data, { id: data.length + 1, todo: addValue }]);
+
+  const handleAdd = () => {
+    setData([
+      ...data,
+      { id: Math.floor(Math.random() * 1000000), todo: addValue },
```
Comment (yasserhennawi) [inline src/App.js]: Can we create a separate util function called `getRandomNumber` that: 
  - take argument called `digits` and raise 10 to the power of this argument
  - returns a random number with `digits` number of digits.
  
  for example: 
    `getRandomNumber(5)` could return `12345`

File: src/App.js [hunk 2]
```diff
@@ -3,57 +3,57 @@ import { useEffect, useState } from "react";
...
+    ]);
     setAddValue("");
   };
 
-  const handleEdite = (id) => {
-    setIsEdit("m-3");
-    setEditId(id);
+  const handleEdite = (index) => {
```
Comment (yasserhennawi) [inline src/App.js]: handleEdit*

File: src/App.js [hunk 3]
```diff
@@ -3,57 +3,57 @@ import { useEffect, useState } from "react";
...
+    const newData = data.slice();
```
Comment (yasserhennawi) [inline src/App.js]: [Optional] Let's use spread instead (since it works with objects as well)

File: src/App.js (hunk 1 above)
Comment (yasserhennawi) [reply src/App.js]: This function should be added in a file called `utils.js`

File: src/App.js [hunk 4]
```diff
@@ -70,59 +70,66 @@ const App = () => {
...
+                <div
+                  key={item.id}
+                  className="flex flex-row w-auto justify-between"
+                >
+                  <input
+                    type="text"
+                    value={item.todo}
+                    className="border w-max rounded-2xl"
+                    onChange={(e) => handleChangeOfEdit(e, index)}
+                  />
```
Comment (yasserhennawi) [inline src/App.js]: this input field appears short and hide the text

image

File: src/App.js [hunk 5]
```diff
@@ -70,59 +70,66 @@ const App = () => {
...
+              item.isEditMode ? (
+                <div
+                  key={item.id}
+                  className="flex flex-row w-auto justify-between"
+                >
+                  <input
+                    type="text"
+                    value={item.todo}
+                    className="border w-max rounded-2xl"
+                    onChange={(e) => handleChangeOfEdit(e, index)}
```
Comment (yasserhennawi) [inline src/App.js]: let's rename `handleChangeOfEdit` to `handleEditChange`

File: src/App.js [hunk 6]
```diff
@@ -3,57 +3,57 @@ import { useEffect, useState } from "react";
...
-    const upateDatAfterDelete = data.filter((item) => {
+    const updateDataAfterDelete = data.filter((item) => {
       return item.id !== id;
     });
-    console.log("upateDatAfterDelete: ", upateDatAfterDelete);
-    setData(upateDatAfterDelete);
+    setData(updateDataAfterDelete);
   };
+
   const handleInputChange = (e) => {
```
Comment (yasserhennawi) [inline src/App.js]: let's rename `handleInputChange` to `handleAddChange`

File: src/App.js [hunk 7]
```diff
@@ -3,57 +3,57 @@ import { useEffect, useState } from "react";
...
-    itemToEdit
-      ? (itemToEdit.todo = valueEdit)
-      : (itemToEdit.todo = itemToEdit.todo);
-    setData(editedArry);
-    setEditValue("");
-    setIsEdit("hidden block");
+  const handleFinshEdit = (index) => {
+    const newData = data.slice();
+    newData[index].isEditMode = false;
+    setData(newData);
```
Comment (yasserhennawi) [inline src/App.js]: Replace `handleEdite` and `handleFinshEdit` with `toggleEditMode` to switch each todo from `editMode = true` to `false`

File: src/App.js [hunk 8]
```diff
@@ -3,57 +3,57 @@ import { useEffect, useState } from "react";
...
 
   useEffect(() => {
     const getData = async () => {
       const response = await fetch("https://dummyjson.com/todos").then((res) =>
         res.json()
       );
-
-      const newArry = [...data, ...response.todos];
-      // console.log(newArry);
+      const newArry = [
```
Comment (yasserhennawi) [inline src/App.js]: newArray*

File: src/App.js [hunk 9]
```diff
@@ -70,59 +70,66 @@ const App = () => {
           className="border w-96 h-20"
         />
```
Comment (yasserhennawi) [inline src/App.js]: Edit "Add todo" input height.

File: src/App.js [hunk 10]
```diff
@@ -70,59 +70,66 @@ const App = () => {
...
         <button
-          className="btn btn-primary ml-3"
+          className={`btn btn-primary ml-3 ${
+            addValue.trim() === "" ? "disabled" : ""
+          }`}
```
Comment (yasserhennawi) [inline src/App.js]: Let's use [classname](https://www.npmjs.com/package/classnames) package.

File: src/App.js [hunk 11]
```diff
@@ -70,59 +70,66 @@ const App = () => {
...
+                    value={item.todo}
+                    className="border w-max rounded-2xl"
+                    onChange={(e) => handleChangeOfEdit(e, index)}
+                  />
+                  <div className="w-auto mr-20">
+                    <button
+                      onClick={() => handleFinshEdit(index)}
+                      className="btn btn-primary m-2"
+                    >
+                      OK
```
Comment (yasserhennawi) [inline src/App.js]: Let's rename this to "FINISH"

File: src/App.js [hunk 12]
```diff
@@ -70,59 +70,66 @@ const App = () => {
...
+                    <button
+                      className="btn btn-primary m-2"
+                      onClick={() => {
+                        handleDelete(item.id);
+                      }}
+                    >
+                      DELETE
+                    </button>
+                    <button
+                      onClick={() => handleEdite(index)}
```
Comment (yasserhennawi) [inline src/App.js]: Migrate the edit logic from using "index" to use the "id"

**Review by yasserhennawi — CHANGES_REQUESTED** [review]
Awesome work 💪 🚀 ✨

**Comment by yasserhennawi** [comment]
~Disabled "Finish edit" button if the field is empty or has "spaces" only.~

File: src/App.js (hunk 1 above)
Comment (Mahmoud-AbdElaziz-Hamed) [reply src/App.js]: done

File: src/App.js (hunk 2 above)
Comment (Mahmoud-AbdElaziz-Hamed) [reply src/App.js]: done

File: src/App.js (hunk 3 above)
Comment (Mahmoud-AbdElaziz-Hamed) [reply src/App.js]: done

File: src/App.js (hunk 4 above)
Comment (Mahmoud-AbdElaziz-Hamed) [reply src/App.js]: done

File: src/App.js (hunk 5 above)
Comment (Mahmoud-AbdElaziz-Hamed) [reply src/App.js]: done

File: src/App.js (hunk 8 above)
Comment (Mahmoud-AbdElaziz-Hamed) [reply src/App.js]: done

File: src/App.js (hunk 9 above)
Comment (Mahmoud-AbdElaziz-Hamed) [reply src/App.js]: done

File: src/App.js (hunk 11 above)
Comment (Mahmoud-AbdElaziz-Hamed) [reply src/App.js]: done

File: src/App.js (hunk 12 above)
Comment (Mahmoud-AbdElaziz-Hamed) [reply src/App.js]: done

File: src/App.js (hunk 10 above)
Comment (Mahmoud-AbdElaziz-Hamed) [reply src/App.js]: done , i have used it in addTodo File in condition adding button

File: src/App.js (hunk 6 above)
Comment (Mahmoud-AbdElaziz-Hamed) [reply src/App.js]: done

File: src/App.js (hunk 7 above)
Comment (Mahmoud-AbdElaziz-Hamed) [reply src/App.js]: done

File: src/App.js (hunk 1 above)
Comment (yasserhennawi) [reply src/App.js]: Super awesome 🎉  
Now let's make sure that it's not **Less than** the passed argument.

Comment (yasserhennawi) [reply src/App.js]: Also naming should be camelCase (For any utility function)

**Comment by yasserhennawi** [comment]
Components and containers folders should have the same ReactComponent name, for example: 

- `containers/Homepage/index.js` has `Homepage` component
- `components/Button/index.js` has `Button` component

**Comment by yasserhennawi** [comment]
PascalCase => React fucntional component
camelCase => normal functions

**Comment by yasserhennawi** [comment]
Please change any `Eidt` to `Edit`

**Comment by yasserhennawi** [comment]
Rename "ok" to "Finish edit" and align it to the end (meaning: it should be like the "Edit" button)


image

**Comment by yasserhennawi** [comment]
Please remove any **trailing space** inside classnames (or string in general)

**Comment by yasserhennawi** [comment]
'disabled' classname (with the condition `value.trim() === ""`) can be merged with the previous class.

**Comment by yasserhennawi** [comment]
Instead of passing `onClick` attribute with `undefined` value, we should use the `disabled` attribute.

File: src/App.js (hunk 11 above)
Comment (yasserhennawi) [reply src/App.js]: Please revisit.

File: src/App.js (hunk 12 above)
Comment (yasserhennawi) [reply src/App.js]: Please review.
Currently it's still the index (renamed Id)

**Comment by yasserhennawi** [comment]
onChange attribute should take prop "onBlaBlaChange"
onClick attribute should take prop "onBlaBlaClick"

**Comment by yasserhennawi** [comment]
replace all export default or named exports 
of example: 
`export default Button` and `import Button from './Button'`
should be 
`export const Button` and `import { Button } from './Button'`

In all react component and util function

**Comment by yasserhennawi** [comment]
in `TodoItem` component, remove `itemTodo` and use `value` instead.

**Comment by yasserhennawi** [comment]
In button component, change `text` prop to be `children` then pass it value as "children" not as prop.

**Comment by yasserhennawi** [comment]
remove any console.log before pushing

**Comment by yasserhennawi** [comment]
Sho3'l gamed fash5 <3 🚀

File: src/App.js (hunk 1 above)
Comment (Mahmoud-AbdElaziz-Hamed) [reply src/App.js]: done

Comment (yasserhennawi) [reply src/App.js]: Please revisit, the function still returns 5digits number sometimes.

**Comment by yasserhennawi** [comment]
Please review onChange & OnClick for all functions in TodoItem and TodoItemList

**Comment by yasserhennawi** [comment]
Replace `toggleEditMode` logic to use `.find()`

**Comment by yasserhennawi** [comment]
`handleEditChange` and `toggleEditMode` should only use `id` as argument.

**Comment by yasserhennawi** [comment]
Disabled behavior for the button should be in the `Button` component
Not TodoItem, Not AddTodo

File: src/App.js (hunk 1 above)
Comment (Mahmoud-AbdElaziz-Hamed) [reply src/App.js]: done

File: src/utils/getRandomNumber/index.test.js [hunk 13]
```diff
@@ -0,0 +1,9 @@
+import { getRandomNumber } from "./index";
+
+describe("getRandomNumber function", () => {
+  test("generates a 6-digit number", () => {
+    const digit = 6;
+    const randomNumber = getRandomNumber(digit);
+    expect(randomNumber.toString().length).toBe(6);
```
Comment (yasserhennawi) [inline src/utils/getRandomNumber/index.test.js:7]: we should here use `digit` instead of `6` directly.

Comment (Mahmoud-AbdElaziz-Hamed) [reply src/utils/getRandomNumber/index.test.js:7]: done

File: src/containers/homePage/index.js [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+  };
+
+  const toggleEditMode = (id) => {
+    const newData = [...data];
+    const editedElement = newData.find((ele) => ele.id === id);
+    editedElement.isEditMode = !editedElement.isEditMode;
+    setData(newData);
+  };
+
+  const handleEditChange = (e, id, todo) => {
```
Comment (yasserhennawi) [inline src/containers/homePage/index.js]: This should only take 2 argument (e, id)

**Comment by yasserhennawi** [comment]
Remove unused customKey prop.

File: src/utils/getRandomNumber/index.js [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+
+//   return randomNumbersArray;
+// };
+
+// // Example usage
+// const digit = 6; // Set the desired number of digits
+// const count = 1000; // Set the desired count of random numbers
+
+// const arrayOfRandomNumbers = generateRandomNumbersArray(digit, count);
+// console.log(arrayOfRandomNumbers);
```
Comment (yasserhennawi) [inline src/utils/getRandomNumber/index.js]: Remove unused comments.

**Review by yasserhennawi — APPROVED** [review]

File: src/containers/homePage/index.js (hunk 14 above)
Comment (Mahmoud-AbdElaziz-Hamed) [reply src/containers/homePage/index.js]: done

File: src/utils/getRandomNumber/index.js (hunk 15 above)
Comment (Mahmoud-AbdElaziz-Hamed) [reply src/utils/getRandomNumber/index.js]: done

**Comment by Mahmoud-AbdElaziz-Hamed** [comment]
@AhmedAbdElaziz-dev  please review the code when you are not busy

**Comment by yasserhennawi** [comment]
✅  Very good PR 🥇 🚀

File: src/components/button/index.js [hunk 16]
```diff
@@ -0,0 +1,13 @@
+export const Button = ({ onClick, children, disabled }) => {
+  return (
+    <button
+      onClick={onClick}
+      className={`text-white py-2 px-4 rounded m-2 ${
+        disabled === true ? "bg-gray-400" : "bg-blue-500"
```
Comment (AhmedAbdElaziz-dev) [inline src/components/button/index.js]: no need for disabled === true as it is boolean you can use it direct

Comment (Mahmoud-AbdElaziz-Hamed) [reply src/components/button/index.js]: done

**Pull request merged**